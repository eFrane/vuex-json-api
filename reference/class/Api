<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Json:Api for Vuex</title>
    <meta name="generator" content="VuePress 1.8.0">
    
    <meta name="description" content="Automagic Json:Api integration for Vuex">
    
    <link rel="preload" href="/assets/css/0.styles.ba26daf8.css" as="style"><link rel="preload" href="/assets/js/app.bbfb528b.js" as="script"><link rel="preload" href="/assets/js/2.0dafffee.js" as="script"><link rel="preload" href="/assets/js/25.4f939962.js" as="script"><link rel="prefetch" href="/assets/js/10.6f689e45.js"><link rel="prefetch" href="/assets/js/11.3e854a07.js"><link rel="prefetch" href="/assets/js/12.5e6fc63d.js"><link rel="prefetch" href="/assets/js/13.b6b5cdaa.js"><link rel="prefetch" href="/assets/js/14.c2149562.js"><link rel="prefetch" href="/assets/js/15.3b44d0e0.js"><link rel="prefetch" href="/assets/js/16.5963f83e.js"><link rel="prefetch" href="/assets/js/17.e10ac304.js"><link rel="prefetch" href="/assets/js/18.557daba9.js"><link rel="prefetch" href="/assets/js/19.d48ac0dd.js"><link rel="prefetch" href="/assets/js/20.f830c59d.js"><link rel="prefetch" href="/assets/js/21.47a4ed11.js"><link rel="prefetch" href="/assets/js/22.cb5d3132.js"><link rel="prefetch" href="/assets/js/23.7c076f99.js"><link rel="prefetch" href="/assets/js/24.6ac270b7.js"><link rel="prefetch" href="/assets/js/26.0bf0a7fd.js"><link rel="prefetch" href="/assets/js/27.99d4a990.js"><link rel="prefetch" href="/assets/js/28.cdd1dd4b.js"><link rel="prefetch" href="/assets/js/29.4e9281f4.js"><link rel="prefetch" href="/assets/js/3.63095073.js"><link rel="prefetch" href="/assets/js/30.fbd48ebc.js"><link rel="prefetch" href="/assets/js/31.67d4c9e9.js"><link rel="prefetch" href="/assets/js/32.aa9038ac.js"><link rel="prefetch" href="/assets/js/33.228977a9.js"><link rel="prefetch" href="/assets/js/34.8be564c9.js"><link rel="prefetch" href="/assets/js/35.6b3da8ae.js"><link rel="prefetch" href="/assets/js/36.329aa466.js"><link rel="prefetch" href="/assets/js/37.00eac06c.js"><link rel="prefetch" href="/assets/js/38.aed504b0.js"><link rel="prefetch" href="/assets/js/39.1223306c.js"><link rel="prefetch" href="/assets/js/4.bfe88385.js"><link rel="prefetch" href="/assets/js/40.04386673.js"><link rel="prefetch" href="/assets/js/41.15514872.js"><link rel="prefetch" href="/assets/js/42.ac8985e6.js"><link rel="prefetch" href="/assets/js/43.75e46776.js"><link rel="prefetch" href="/assets/js/44.f73cfb01.js"><link rel="prefetch" href="/assets/js/45.e2c42640.js"><link rel="prefetch" href="/assets/js/46.8c55b03e.js"><link rel="prefetch" href="/assets/js/47.f35fbef8.js"><link rel="prefetch" href="/assets/js/48.8b580e70.js"><link rel="prefetch" href="/assets/js/49.57c4a9c9.js"><link rel="prefetch" href="/assets/js/5.cb249b63.js"><link rel="prefetch" href="/assets/js/50.3fd1e7cb.js"><link rel="prefetch" href="/assets/js/51.4c6b4362.js"><link rel="prefetch" href="/assets/js/52.264f1842.js"><link rel="prefetch" href="/assets/js/53.49f54f43.js"><link rel="prefetch" href="/assets/js/54.4a3a9634.js"><link rel="prefetch" href="/assets/js/55.933fb422.js"><link rel="prefetch" href="/assets/js/56.f2e73cb2.js"><link rel="prefetch" href="/assets/js/57.b9231f84.js"><link rel="prefetch" href="/assets/js/58.83ee0600.js"><link rel="prefetch" href="/assets/js/59.adde428b.js"><link rel="prefetch" href="/assets/js/6.4155846f.js"><link rel="prefetch" href="/assets/js/60.e3dfd946.js"><link rel="prefetch" href="/assets/js/61.f44f8aa3.js"><link rel="prefetch" href="/assets/js/62.1a8322f7.js"><link rel="prefetch" href="/assets/js/63.41a5a998.js"><link rel="prefetch" href="/assets/js/64.7c1cb518.js"><link rel="prefetch" href="/assets/js/65.b037b74d.js"><link rel="prefetch" href="/assets/js/66.a4187ea3.js"><link rel="prefetch" href="/assets/js/67.95fb4f5e.js"><link rel="prefetch" href="/assets/js/7.6ed4aaa9.js"><link rel="prefetch" href="/assets/js/8.58c932af.js"><link rel="prefetch" href="/assets/js/9.9a6f2cf9.js">
    <link rel="stylesheet" href="/assets/css/0.styles.ba26daf8.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container no-sidebar"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"><!----> <span class="site-name">Json:Api for Vuex</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="/" class="nav-link">
  Home
</a></div><div class="nav-item"><a href="/reference/" class="nav-link router-link-active">
  Reference
</a></div><div class="nav-item"><a href="/usage/" class="nav-link">
  Usage
</a></div> <a href="https://github.com/eFrane/vuex-json-api" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/" class="nav-link">
  Home
</a></div><div class="nav-item"><a href="/reference/" class="nav-link router-link-active">
  Reference
</a></div><div class="nav-item"><a href="/usage/" class="nav-link">
  Usage
</a></div> <a href="https://github.com/eFrane/vuex-json-api" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav>  <!----> </aside> <main class="page"> <div class="theme-default-content content__default"><p><a name="Api"></a></p> <h2 id="api"><a href="#api" class="header-anchor">#</a> Api</h2> <p><strong>Kind</strong>: global class</p> <ul><li><a href="#Api">Api</a> <ul><li><a href="#new_Api_new">new Api()</a></li> <li><a href="#Api+setPreprocessingCallbacks">.setPreprocessingCallbacks(callbacks)</a></li> <li><a href="#Api+setErrorCallbacks">.setErrorCallbacks(callbacks)</a></li> <li><a href="#Api+addPreprocessingCallback">.addPreprocessingCallback(callback)</a></li> <li><a href="#Api+addErrorCallback">.addErrorCallback(callback)</a></li> <li><a href="#Api+resetPreprocessing">.resetPreprocessing()</a></li> <li><a href="#Api+resetErrorCallbacks">.resetErrorCallbacks()</a></li> <li><a href="#Api+addHeaders">.addHeaders(headers)</a></li></ul></li></ul> <p><a name="new_Api_new"></a></p> <h3 id="new-api"><a href="#new-api" class="header-anchor">#</a> new Api()</h3> <p>Wrapper around json:api requests, sets content type and other defaults.</p> <p><a name="Api+setPreprocessingCallbacks"></a></p> <h3 id="api-setpreprocessingcallbacks-callbacks"><a href="#api-setpreprocessingcallbacks-callbacks" class="header-anchor">#</a> api.setPreprocessingCallbacks(callbacks)</h3> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>callbacks</td> <td><code>Array</code></td></tr></tbody></table> <p><a name="Api+setErrorCallbacks"></a></p> <h3 id="api-seterrorcallbacks-callbacks"><a href="#api-seterrorcallbacks-callbacks" class="header-anchor">#</a> api.setErrorCallbacks(callbacks)</h3> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>callbacks</td> <td><code>Array</code></td></tr></tbody></table> <p><a name="Api+addPreprocessingCallback"></a></p> <h3 id="api-addpreprocessingcallback-callback"><a href="#api-addpreprocessingcallback-callback" class="header-anchor">#</a> api.addPreprocessingCallback(callback)</h3> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>callback</td> <td><code>function</code></td></tr></tbody></table> <p><a name="Api+addErrorCallback"></a></p> <h3 id="api-adderrorcallback-callback"><a href="#api-adderrorcallback-callback" class="header-anchor">#</a> api.addErrorCallback(callback)</h3> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>callback</td> <td><code>function</code></td></tr></tbody></table> <p><a name="Api+resetPreprocessing"></a></p> <h3 id="api-resetpreprocessing"><a href="#api-resetpreprocessing" class="header-anchor">#</a> api.resetPreprocessing()</h3> <p>Reset the response preprocessing to the default behaviour</p> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a><br> <a name="Api+resetErrorCallbacks"></a></p> <h3 id="api-reseterrorcallbacks"><a href="#api-reseterrorcallbacks" class="header-anchor">#</a> api.resetErrorCallbacks()</h3> <p>Reset the response error to the default behaviour</p> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a><br> <a name="Api+addHeaders"></a></p> <h3 id="api-addheaders-headers"><a href="#api-addheaders-headers" class="header-anchor">#</a> api.addHeaders(headers)</h3> <p><strong>Kind</strong>: instance method of <a href="#Api"><code>Api</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>headers</td> <td><code>Object</code></td></tr></tbody></table></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/reference/" class="prev router-link-active">
        Reference (JS)
      </a></span> <!----></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.bbfb528b.js" defer></script><script src="/assets/js/2.0dafffee.js" defer></script><script src="/assets/js/25.4f939962.js" defer></script>
  </body>
</html>
