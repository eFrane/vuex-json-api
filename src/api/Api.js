import axios from 'axios'
import { stringify } from 'qs'
import { validateCallbackFn } from '../helpers/validateCallbackFn'
import { setCallbackFns } from '../helpers/setCallbackFns'
import { isAbsoluteUri } from '../helpers/isAbsoluteUri'

class Api {
  constructor () {
    this.baseUrl = ''

    this.preprocessingCallbacks = []
    this.errorCallbacks = []

    this.defaultOptions = {
      paramsSerializer (params) {
        return stringify(params, { encodeValuesOnly: true, arrayFormat: 'brackets' })
      }
    }

    this.headers = {
      'X-Requested-With': 'XMLHttpRequest',
      'Content-Type': 'application/vnd.api+json',
      Accept: 'application/vnd.api+json'
    }
  }

  /**
   *
   * @param {Array} callbacks
   */
  setPreprocessingCallbacks (callbacks) {
    this.preprocessingCallbacks = setCallbackFns(callbacks)
  }

  /**
   *
   * @param {Array} callbacks
   */
  setErrorCallbacks (callbacks) {
    this.errorCallbacks = setCallbackFns(callbacks)
  }

  /**
   *
   * @param {Function} callback
   */
  addPreprocessingCallback (callback) {
    if (validateCallbackFn(callback)) {
      throw new Error('You must pass a valid callback to this method')
    }

    this.preprocessingCallbacks.push(callback)
  }

  /**
   *
   * @param {Function} callback
   */
  addErrorCallback (callback) {
    if (validateCallbackFn(callback)) {
      throw new Error('You must pass a valid callback to this method')
    }

    this.errorCallbacks.push(callback)
  }

  /**
   * Reset the response preprocessing to the default behaviour
   */
  resetPreprocessing () {
    this.preprocessingCallbacks = []
  }

  /**
   * Reset the response error to the default behaviour
   */
  resetErrorCallbacks () {
    this.errorCallbacks = []
  }

  /**
   *
   * @param {Object} headers
   */
  addHeaders (headers) {
    if (headers.constructor !== Object) {
      throw new Error('You must pass an object to this method')
    }

    // make sure there is a spreadable object in headers
    if (!headers || Object.getOwnPropertyNames(headers).length < 1) {
      headers = {}
    }

    this.headers = Object.assign(this.headers, headers)
  }

  async doRequest (method, url, params, data, options) {
    const config = Object.assign(options, this.defaultOptions)
    config.headers = this.headers

    if (!isAbsoluteUri(url)) {
      url = this.baseUrl + url
    }

    // make cross domain requests if necessary
    let crossDomain = false
    if (url.length > 0 && isAbsoluteUri(url)) {
      crossDomain = true
    }

    return axios.create(config)
      .request({ method, url, params, data, crossDomain })
      .then(
        async response => {
          for (let i = 0; i < this.preprocessingCallbacks.length; i++) {
            await this.preprocessingCallbacks[i](response)
          }

          return response
        }
      ).catch(
        async errorResponse => {
          for (let i = 0; i < this.errorCallbacks.length; i++) {
            await this.errorCallbacks[i](errorResponse)
          }

          return errorResponse
        }
      )
  }

  get (url, params = null, options = {}) {
    return this.doRequest('get', url, params, null, options)
  }

  post (url, params = null, data = null, options = {}) {
    return this.doRequest('post', url, params, data, options)
  }

  put (url, params = null, data = null, options = {}) {
    return this.doRequest('put', url, params, data, options)
  }

  patch (url, params = null, data = null, options = {}) {
    return this.doRequest('patch', url, params, data, options)
  }

  delete (url, params = null, data = null, options = {}) {
    return this.doRequest('delete', url, params, data, options)
  }

  setBaseUrl (baseUrl) {
    this.baseUrl = baseUrl
  }
}

export { Api }
