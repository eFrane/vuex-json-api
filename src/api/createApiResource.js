export function createApiResource () {
  console.time('api: add method proxies for route ' + routeName)

  for (let methodName in methods) {
    if (methods.hasOwnProperty(methodName)) {
      let route = methods[methodName]

      if (methodName.indexOf('related.') === 0) {
        this.registerRelatedResourceMethod(routeName, methodName, route)
        continue
      }

      this[routeName][methodName] = createApiResourceMethodProxy(this, methodName, route)
    }
  }

  console.timeEnd('api: add method proxies for route ' + routeName)
}
