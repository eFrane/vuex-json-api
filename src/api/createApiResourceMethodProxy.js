import { Route } from '../route/Route'

/**
*
* @param {Api} api
* @param {String} method
* @param {Route} route
*/
export function createApiResourceMethodProxy (api, method, route) {
  return new Proxy(() => {}, {
    apply (target, thisArg, argArray) {
      if (!(route instanceof Route)) {
        throw new Error('Expected Route object')
      }

      // add actual route as first param
      let url = route.prepare(argArray[0])
      argArray.unshift(url)

      switch (method) {
        case 'list':
        case 'get':
          return api.get.apply(api, argArray)

        case 'create':
          return api.post.apply(api, argArray)

        case 'replace':
          return api.put.apply(api, argArray)

        case 'update':
          return api.patch.apply(api, argArray)

        case 'delete':
          return api.delete.apply(api, argArray)

        default:
          throw new Error('unsupported api method: ' + method)
      }
    }
  })
}
