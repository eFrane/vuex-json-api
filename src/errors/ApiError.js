export class ApiError extends Error {
  constructor (message) {
    super(message)

    this.name = 'ApiError'
  }
}

export class NotFoundApiError extends ApiError {
  constructor (message, errorInfo = null) {
    super(message)

    this.name = 'NotFoundApiError'
    this.errorInfo = errorInfo
  }

  hasErrorInfo () {
    return this.errorInfo !== null && this.errorInfo !== undefined
  }
}
