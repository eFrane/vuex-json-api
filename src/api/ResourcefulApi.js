import normalize from 'json-api-normalizer'

import { Api } from './Api'
import { ModuleBuilder } from '../module/ModuleBuilder'

export class ResourcefulApi extends Api {
  async doRequest (method, url, params, data, options) {
    return super.doRequest(method, url, params, data, options)
      .then((response) => {
        return {
          data: normalize(response.data),
          meta: response.data.meta,
          status: response.status
        }
      })
  }

  /**
   *
   * @param {route.Router} router
   */
  setupResourcefulRequests (router) {
    this.router = router

    console.time('api: setup resourceful routing')
    const routes = router.getRoutes()
    this.registerableModules = {}

    for (const routeName in routes) {
      if (routes.hasOwnProperty(routeName)) {
        let methods = routes[routeName]

        this.registerResourceMethods(routeName, methods)
        this.registerableModules[routeName] = methods
      }
    }

    console.timeEnd('api: setup resourceful routing')
  }

  /**
   *
   * @param {Vuex.Store} store
   * @param {Array} modulesToRegister
   */
  setupModules (store, modulesToRegister = []) {
    console.time('api: setup modules')

    let registerableModuleNames = Object.keys(this.registerableModules)

    let currentModuleName
    do {
      currentModuleName = registerableModuleNames.pop()

      if (typeof currentModuleName !== 'undefined' &&
        (modulesToRegister.length === 0 || modulesToRegister.indexOf(currentModuleName))) {
        this.registerModule(store, this[currentModuleName], currentModuleName)
      }
    } while (currentModuleName)

    delete this.registerableModules

    console.timeEnd('api: setup modules')
  }

  /**
   *
   * @param {Vuex} store
   * @param {Route} methods
   */
  registerModule (store, methods, moduleName) {
    // prevent double registration
    if (store.state.hasOwnProperty(moduleName)) {
      return
    }

    let moduleBuilder = new ModuleBuilder(store, this, moduleName, methods)
    const module = moduleBuilder.build()
    if (moduleName) {
      store.registerModule(moduleName, module)
    }
  }

  /**
   *
   * @param {String} routeName
   * @param {Route} methods
   */
  registerResourceMethods (routeName, methods) {
    this[routeName] = {}

    console.time('api: add method proxies for route ' + routeName)

    for (let methodName in methods) {
      if (methods.hasOwnProperty(methodName)) {
        let route = methods[methodName]

        if (methodName.indexOf('related.') === 0) {
          this.registerRelatedResourceMethod(routeName, methodName, route)
          continue
        }

        this[routeName][methodName] = ResourcefulApi.createApiProxy(this, methodName, route)
      }
    }

    console.timeEnd('api: add method proxies for route ' + routeName)
  }

  registerRelatedResourceMethod (routeName, methodName, route) {
    let [related, resource, relationMethod] = methodName.split('.')

    if (typeof this[routeName][related] !== 'object') {
      this[routeName][related] = {}
    }

    if (typeof this[routeName][related][resource] !== 'object') {
      this[routeName][related][resource] = {}
    }

    this[routeName][related][resource][relationMethod] =
      ResourcefulApi.createApiProxy(this, relationMethod, route)
  }

  /**
   *
   * @param {Api} api
   * @param {String} method
   * @param {Route} route
   */
  static createApiProxy (api, method, route) {
    return new Proxy(() => {}, {
      apply (target, thisArg, argArray) {
        // add actual route as first param
        argArray.unshift(route)

        switch (method) {
          case 'list':
          case 'get':
            return api.get.apply(api, argArray)

          case 'create':
            return api.post.apply(api, argArray)

          case 'replace':
            return api.put.apply(api, argArray)

          case 'update':
            return api.patch.apply(api, argArray)

          case 'delete':
            return api.delete.apply(api, argArray)

          default:
            throw new Error('unsupported api method: ' + method)
        }
      }
    })
  }
}
