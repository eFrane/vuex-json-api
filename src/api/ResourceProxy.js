import { Route } from '../route/Route'
import { createApiResourceMethodProxy } from './createApiResourceMethodProxy'

/**
 * ResourceProxy
 *
 * A wrapper class for the available methods on
 * a route. The actual proxied axios-clients
 * will only be created once the first request on
 * a method is made.
 */
export class ResourceProxy {
  constructor (resourcefulApi) {
    this.resourcefulApi = resourcefulApi
    this.routes = {}
    this.proxies = {}
    this.related = {}
  }

  /**
   * Adds a method to the resource
   *
   * @param {Route} route
   */
  addRoute (route) {
    if (!(route instanceof Route)) {
      throw new Error(`Expected Route, got ${route.constructor}`)
    }

    this.routes[route.action] = route
  }

  addRelation (relationName, relatedResourceProxy) {
    if (!(relatedResourceProxy instanceof ResourceProxy)) {
      throw new Error(`Expected ResourceProxy, got ${relatedResourceProxy.constructor}`)
    }

    this.related[relationName] = relatedResourceProxy
  }

  /**
   * Used by the resource method availability checks
   * to indicate a missing method.
   *
   * @param {string} methodName
   */
  methodNotAvailable (methodName) {
    throw new Error(`Method ${methodName} not available`)
  }

  getProxyForMethod (methodName) {
    if (!Object.prototype.hasOwnProperty.call(this.proxies, methodName)) {
      this.createProxyForMethodIfMissing(methodName)
    }

    return this.proxies[methodName]
  }

  createProxyForMethodIfMissing (methodName) {
    const proxy = createApiResourceMethodProxy(this.api, methodName, this.routes[methodName])

    this.proxies[methodName] = proxy
  }

  get () {
    return ('get' in this.routes) ? this.getProxyForMethod('get') : this.methodNotAvailable('get')
  }

  list () {
    return ('list' in this.routes) ? this.getProxyForMethod('list') : this.methodNotAvailable('list')
  }

  create () {
    return ('create' in this.routes) ? this.getProxyForMethod('create') : this.methodNotAvailable('create')
  }

  replace () {
    return ('replace' in this.routes) ? this.getProxyForMethod('replace') : this.methodNotAvailable('replace')
  }

  update () {
    return ('update' in this.routes) ? this.getProxyForMethod('update') : this.methodNotAvailable('update')
  }

  delete () {
    return ('delete' in this.routes) ? this.getProxyForMethod('delete') : this.methodNotAvailable('delete')
  }
}
