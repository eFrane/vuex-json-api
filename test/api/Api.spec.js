/**
 * @jest-environment jsdom
 */

import { Api } from '@/api/Api'
import { ApiError } from '@/errors/ApiError'

// This spec tests all the non-request methods of Api

const api = new Api()
api.setBaseUrl('http://localhost/')

test('has no default base url', () => {
  expect(Api.baseUrl).toBeUndefined()

  const unconfiguredApi = new Api()

  expect(unconfiguredApi.baseUrl).toBe('')
})

test('allows changing the base url', () => {
  const unconfiguredApi = new Api()
  unconfiguredApi.setBaseUrl('/api/')

  expect(unconfiguredApi.baseUrl).toBe('/api/')
})

test('setting callbacks', () => {
  const validCb = () => {}
  const invalidCb = 42

  expect(api.preprocessingCallbacks.length).toBe(0)
  expect(api.errorCallbacks.length).toBe(0)

  expect(() => {
    api.addPreprocessingCallback(validCb)
  }).not.toThrow(ApiError)

  expect(() => {
    api.addPreprocessingCallback(invalidCb)
  }).toThrow(ApiError)

  expect(() => {
    api.addErrorCallback(validCb)
  }).not.toThrow(ApiError)

  expect(() => {
    api.addErrorCallback(invalidCb)
  }).toThrow(ApiError)

  expect(api.preprocessingCallbacks.length).toBe(1)
  expect(api.errorCallbacks.length).toBe(1)

  api.resetPreprocessing()
  api.resetErrorCallbacks()

  expect(api.preprocessingCallbacks.length).toBe(0)
  expect(api.errorCallbacks.length).toBe(0)

  expect(() => {
    api.setPreprocessingCallbacks([validCb, invalidCb])
  }).toThrow(ApiError)

  expect(api.preprocessingCallbacks.length).toBe(1)

  expect(() => {
    api.setPreprocessingCallbacks([invalidCb, validCb])
  }).toThrow(ApiError)

  expect(api.preprocessingCallbacks.length).toBe(2)

  expect(() => {
    api.setPreprocessingCallbacks([validCb])
  }).not.toThrow(ApiError)

  expect(api.preprocessingCallbacks.length).toBe(3)

  api.resetPreprocessing()

  expect(() => {
    api.setErrorCallbacks([validCb, invalidCb])
  }).toThrow(ApiError)

  expect(api.errorCallbacks.length).toBe(1)

  expect(() => {
    api.setErrorCallbacks([invalidCb, validCb])
  }).toThrow(ApiError)

  expect(api.errorCallbacks.length).toBe(2)

  expect(() => {
    api.setErrorCallbacks([validCb])
  }).not.toThrow(ApiError)

  expect(api.errorCallbacks.length).toBe(3)

  api.resetErrorCallbacks()
})

test('adding headers', () => {
  expect(api.setHeader).toThrow(ApiError)

  expect(() => {
    api.setHeader(42, 34)
  }).toThrow(ApiError)

  expect(Object.keys(api.headers).length).toBe(2)

  api.setHeader('foo', 'bar')

  expect(Object.keys(api.headers).length).toBe(3)
  expect(api.headers.foo).toBe('bar')

  expect(() => {
    api.setHeader('foo', 'baz')
  }).toThrow(ApiError)

  expect(Object.keys(api.headers).length).toBe(3)

  api.setHeader('foo', 'baz', true)

  expect(api.headers.foo).toBe('baz')
  expect(Object.keys(api.headers).length).toBe(3)

  api.addHeaders({
    extra1: 'a',
    extra2: 'b'
  })

  expect(Object.keys(api.headers).length).toBe(5)
  expect(api.headers.extra2).toBe('b')
})

test('default headers are read-only', () => {
  expect(() => {
    api.setHeader('Accept', 'text/plain')
  }).toThrow(ApiError)
})

describe.each([
  {
    url: '//localhost/',
    cross: false
  },
  {
    url: 'http://localhost/',
    cross: false
  },
  {
    url: '/relative',
    cross: false
  },
  {
    url: 'http://otherdomain.com',
    cross: true
  },
  {
    url: 'https://ssldomain.com',
    cross: true
  },
  {
    url: '//ffoobbaar.net',
    cross: true
  },
  // this test case could be left out with typescript
  {
    url: 42,
    cross: false
  }
])('cross domain tests', (data) => {
  const shouldCross = (data.cross) ? 'is' : 'is not'

  test(`checks if ${data.url} ${shouldCross} cross domain`, () => {
    expect(api.isUrlCrossDomain(data.url)).toBe(data.cross)
  })
})

describe.each([
  {
    base: 'http://localhost/',
    url: '//foo.com',
    params: {},
    result: 'http://foo.com/'
  },
  {
    base: 'http://localhost/',
    url: '/relative',
    params: {
      foo: true,
      bar: 21
    },
    result: 'http://localhost/relative?foo=true&bar=21'
  },
  {
    base: 'http://localhost/api/',
    url: 'relative',
    params: {},
    result: 'http://localhost/api/relative'
  },
  {
    base: 'http://localhost/api',
    url: 'relative',
    params: {},
    result: 'http://localhost/api/relative'
  },
  {
    base: 'http://localhost/api/',
    url: '/relative',
    params: {},
    result: 'http://localhost/api/relative'
  }
])('url building tests', (data) => {
  const {
    base,
    url,
    params,
    result
  } = data

  test(`combines ${base}, ${url} and ${Object.keys(params).length} parameters to correct url ${result}`, () => {
    api.setBaseUrl(base)
    expect(api._compileUrl(url, params)).toEqual(result)
  })
})
