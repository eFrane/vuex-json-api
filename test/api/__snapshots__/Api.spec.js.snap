// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding headers: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 1`] = `"Expected name to be string, got undefined"`;

exports[`adding headers: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 2`] = `"Expected name to be string, got number"`;

exports[`adding headers: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 3`] = `"Header is already set, not overwriting."`;

exports[`default headers are read-only: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 1`] = `"Cannot change default headers"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 1`] = `"You must pass a valid callback to this method"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 2`] = `"You must pass a valid callback to this method"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 3`] = `"Invalid callbacks: () => {}, 42"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 4`] = `"Invalid callbacks: 42, () => {}"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 5`] = `"Invalid callbacks: () => {}, 42"`;

exports[`setting callbacks: class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ApiError';
  }

} 6`] = `"Invalid callbacks: 42, () => {}"`;
